File: code/python/observer/Teacher.py
class Teacher:
    def __init__(self):
        self.students = []

    def add_student(self, student):
        print(f"Teacher added a student {type(student).__name__}")
        self.students.append(student)

    def remove_student(self, student):
        print(f"Teacher removed a student {type(student).__name__}")
        self.students.remove(student)

    def notify_students(self):
        print("Teacher is notifying students.")
        for student in self.students:
            student.receive_update()

File: code/typescript/src/observer/Teacher.ts
import { Student } from "./Student";

export class Teacher {
    private students: Student[] = [];

    addStudent(student: Student): void {
        console.log(`Teacher added a student ${student.constructor.name}`);
        this.students.push(student);
    }

    removeStudent(student: Student): void {
        console.log(`Teacher removed a student ${student.constructor.name}`);
        const index = this.students.indexOf(student);
        if (index !== -1) {
            this.students.splice(index, 1);
        }
    }

    notifyStudents(): void {
        console.log("Teacher is notifying students.");
        this.students.forEach(student => {
            student.receiveUpdate();
        });
    }
}
